File: merkle-trie/src/main/java/net/consensys/cava/trie/NodeFactory.java
package net.consensys.cava.trie;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.concurrent.AsyncResult;

import java.util.ArrayList;
import java.util.Optional;

interface NodeFactory<V> {

  AsyncResult<Node<V>> createExtension(Bytes path, Node<V> child);

  AsyncResult<Node<V>> createBranch(byte leftIndex, Node<V> left, byte rightIndex, Node<V> right);

  AsyncResult<Node<V>> createBranch(ArrayList<Node<V>> newChildren, Optional<V> value);

  AsyncResult<Node<V>> createLeaf(Bytes path, V value);
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/NodeFactory.kt
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes

internal interface NodeFactory<V> {

  suspend fun createExtension(path: Bytes, child: Node<V>): Node<V>

  suspend fun createBranch(leftIndex: Byte, left: Node<V>, rightIndex: Byte, right: Node<V>): Node<V>

  suspend fun createBranch(newChildren: List<Node<V>>, value: V?): Node<V>

  suspend fun createLeaf(path: Bytes, value: V): Node<V>
}
